src/11_6_2021_RcppClock_benchmarking_Rcpp_code.Rmd

```diff
@@ -0,0 +1,148 @@
+---
+title: "Benchmarking Rcpp code with RcppClock"
+author: "Zach DeBruine"
+license: GPL (>= 2)
+tags: Rcpp benchmarking RcppClock
+summary: Rcpp benchmarking with a simple C++ tick-tock clock and seamless integration
+  with the R environment at runtime.
+---
+
+Rcpp is all about the need for speed. However, Rcpp code is not magically fast, but is the result of careful coding and profiling. RcppClock makes it easy to profile C++ code from R:
+
+* On the C++ side, RcppClock wraps `std::chrono::high_resolution_clock` to time chunks of code, and exports the results to an R object in the global environment.
+* On the R side, RcppClock provides convenient methods for summarizing results and plotting with `ggplot2`.
+
+By benchmarking Rcpp code with RcppClock, we avoid measuring latency associated with converting R types to C++ types (and back again), and we can profile at line-by-line resolution rather than entire functions.
+
+Note: RcppClock is not compatible with OpenMP parallelization, and does not accurately measure processes taking less than ~1 microsecond.
+
+## A simple example
+
+Install RcppClock from CRAN:
+
+```{r, warning = FALSE, message = FALSE, eval = FALSE}
+install.packages("RcppClock")
+```
+
+```{Rcpp, ref.label=knitr::all_rcpp_labels(), include=FALSE}
+```
+
+In your `.cpp` file, link the RcppClock header with `//[[Rcpp::depends(RcppClock)]]` and, if this is an R package, link to it in your `DESCRIPTION` file. `RcppClock` loads `Rcpp.h` so we don't need to include it again.
+
+```{Rcpp, eval = FALSE}
+//[[Rcpp::depends(RcppClock)]]
+#include <RcppClock.h>
+#include <thread>
+
+//[[Rcpp::export]]
+void sleepy(){
+  Rcpp::Clock clock;
+  
+  clock.tick("both_naps");
+  
+  clock.tick("short_nap");
+  std::this_thread::sleep_for(std::chrono::milliseconds(10));  
+  clock.tock("short_nap");
+  
+  clock.tick("long_nap");
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));  
+  clock.tock("long_nap");
+
+  clock.tock("both_naps");
+  
+  // send the times to the R global environment variable, named "naptimes"
+  clock.stop("naptimes");
+}
+```
+
+`.tick(std::string)` starts a new timer. Provide a name to record what is being timed.
+
+`.tock(std::string)` stops a timer. It is important to use the same name as declared in .tick().
+
+`.stop(std::string)` calculates the duration between all .tick() and .tock() timing results, and creates an object in the R environment with the name provided.
+
+On the R end, we can now do stuff with the "naptimes" variable that was created in the above Rcpp function:

```

Comment:
Can you do better in writing style than "do stuff" ?

Reply:
yes indeed.  "do stuff" = "summarize results in R using the 'naptimes' global variable..."